scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala

```diff
@@ -279,24 +240,74 @@ class ScalametaParser(input: Input)(implicit dialect: Dialect) { parser =>
     case Tokenized.Success(tokens) => tokens
     case Tokenized.Error(_, _, details) => throw details
   }
-  lazy val (parserTokens, parserTokenPositions) = {
-    def indentedRegion(regions: List[SepRegion]): Boolean =
-      regions.headOption.exists(_.isIndented)
 
-    val parserTokens = mutable.ArrayBuilder.make[Token]
-    val parserTokenPositions = mutable.ArrayBuilder.make[Int]
+  // NOTE: public methods of TokenIterator return scannerTokens-based positions
+  trait TokenIterator extends Iterator[Token] {
+    def prevTokenPos: Int
+    def tokenPos: Int
+    def token: Token
+    def fork: TokenIterator
+    def observeIndented(): Boolean
+  }
+  var in: TokenIterator = {
+    new NewTokenIterator(
+      scannerTokens,
+      List.empty,
+      TokenRef(scannerTokens(0), 0, 1),
+      -1
+    )
+  }
+
+  case class TokenRef(
+      token: Token,
+      pos: Int,
+      nextPos: Int
+  )
 
-    @tailrec def loop(
+  class NewTokenIterator(
```

Comment:
I have no idea how to name this Iterator 
The previous one was called `SimpleTokenIterator` but this name doesn't fit anymore :)

Is there a need to keep `TokenIterator` interface?

Reply:
`LazyTokenIterator` ?

Reply:
That one is good but what do you think about removing `TokenInterator` interface? Then this name might be used for the current implementation.
Is it used anywhere outside of `ScalametaParser`?